### PR TITLE
fix(buffers): improve disk v2 buffer tests + fix some bugs

### PR DESCRIPTION
**Pull request by tobz**
Quite a few things in this one:
- a lot of cleanup of poorly-written code comments or no-longer-relevant TODOs
- the `buffer_perf` binary has been `clap`ped so now it's way easier to actually use and configure
- fixes related to the finicky nature of not being able to update both the ledger and do a read/write/delete/etc in an atomic fashion, and ensuring disk v2 is more resistant to situations where those torn updates occur

Specifically, in terms of what we've changed in the reader/writer code of the disk buffer v2 code itself..

Prior to this PR, we stored all relevant information in the ledger: total record count, total buffer size, reader/writer positions in terms of file and record ID.  In normal cases, this worked fine and the logic was correct.  However, this all sort of falls apart if the process stops/crashes at just the right moments.  Examples of bad issues:
- record count/buffer size become inconsistent because we updated them after writing but before flushing
- file marked for deletion because the reader has fully processed it, but Vector crashes before we get a chance to delete it

A lot of this is borne out of the design to have the ledger be a very simple progress tracker, essentially,  appear very much like a ring buffer where you literally just know your positions in the buffer, and have little to no concept of anything else.  I still think this was a reasonable design, given our goals, but all of the involved atomics being part of the buffer functioning correctly was at odds with avoiding more formalized designs i.e. having a full-blown write-ahead log or log-structured merge tree design.

To wit, what I've done here is eschewed most of the atomics except for the very core and essential values: the reader and writer file ID positions (which file they're reading from/writing to) and the next writer record ID and last reader record ID.

Knowing which file the reader/writer is in and critical, and can't be avoided.  Since we emulate a ring buffer, this logic is pretty straightforward and can tolerate being interrupted.  The record ID fields, likewise, can also tolerate interruptions:
- the next writer record ID only risks being invalid when we write but don't increment (we increment after a write to avoid potential duplicate record IDs) but this state can be detected
- the last reader record ID functions a lot like normal acknowledgements: if we don't update it correctly, we simply end up rereading those records, so we ensure at-least-once delivery, same as disk buffer v1

We've removed the total record count and total buffer size from the ledger, and derive them directly: total record count from the difference between the reader/writer record ID fields, and total buffer size from an initialization step.

As we have the writer validate the data in the file it left off writing to, as well as have the reader seek to the last record it read in the file it left off reading from, we have a chance during buffer initialization to simply reconstitute that record count/buffer size state directly from the data, and after that happens, we maintain an internal state derived from that.  If Vector crashes, this is OK, because we always reconstitute that internal state: we only need to care about hardening the logic around record IDs and the algorithm of which process creates files vs deletes them, etc.

### CODE REVIEW

File: lib/vector-core/buffers/examples/buffer_perf.rs [hunk 1]
```diff
@@ -1,9 +1,9 @@
 #![allow(clippy::print_stderr)] // soak framework
 #![allow(clippy::print_stdout)]
 // Clippy allows are because this is an example/soak-y test, where we don't
-// actually care about the presence of `println!` calls.
+// actually care about the presence of `info!` calls.
```
Comment (blt) [inline lib/vector-core/buffers/examples/buffer_perf.rs:4]: Regex replace?

File: lib/vector-core/buffers/src/disk_v2/ledger.rs [hunk 2]
```diff
@@ -586,10 +604,57 @@ impl Ledger {
             last_flush: AtomicCell::new(Instant::now()),
             usage_handle,
         };
-        ledger.synchronize_buffer_usage();
+        ledger.update_buffer_size().await?;
 
         Ok(ledger)
     }
+
+    async fn update_buffer_size(&mut self) -> Result<(), LedgerLoadCreateError> {
```
Comment (blt) [inline lib/vector-core/buffers/src/disk_v2/ledger.rs:612]: What happens with this approach if someone starts two vector processes with the same config at the same time?

File: lib/vector-core/buffers/src/disk_v2/reader.rs [hunk 3]
```diff
@@ -457,12 +488,26 @@ where
...
         // implies a data file existing.
         loop {
+            let (reader_file_id, writer_file_id) = self.ledger.get_current_reader_writer_file_id();
             let data_file_path = self.ledger.get_current_reader_data_file_path();
             let data_file = match File::open(&data_file_path).await {
                 Ok(data_file) => data_file,
                 Err(e) => match e.kind() {
                     ErrorKind::NotFound => {
-                        self.ledger.wait_for_writer().await;
+                        // SCREAM SCREAM SCREAM
```
Comment (blt) [inline lib/vector-core/buffers/src/disk_v2/reader.rs:497]: lol

File: lib/vector-core/buffers/src/disk_v2/reader.rs [hunk 4]
```diff
@@ -457,12 +488,26 @@ where
...
+                        // this is a not entirely proven out thought:
+                        // if the writer/reader file IDs are not the same, and we try to open our
+                        // current file and it doesn't exist, what actually happened is that we
+                        // deleted it as instructed but never made it to the "now update our reader
+                        // file ID in the ledger" step.  we should just increment our read file ID
+                        // and try again.
```
Comment (blt) [inline lib/vector-core/buffers/src/disk_v2/reader.rs:503]: I buy this, but it's one of those things I'd _really_ like some property testing to help verify.

**Review by blt — APPROVED** [review]
I buy the approach. I have some questions/comments in the review but nothing that blocks merging this up, as I know it solves an observed set of bugs.

File: lib/vector-core/buffers/src/disk_v2/ledger.rs (hunk 2 above)
Comment (jszwedko) [reply lib/vector-core/buffers/src/disk_v2/ledger.rs:612]: This is a good question. Maybe we should have a lockfile?

File: lib/vector-core/buffers/src/disk_v2/reader.rs (hunk 3 above)
Comment (tobz) [reply lib/vector-core/buffers/src/disk_v2/reader.rs:497]: Woops, forgot to remove that. 😓

File: lib/vector-core/buffers/src/disk_v2/ledger.rs (hunk 2 above)
Comment (tobz) [reply lib/vector-core/buffers/src/disk_v2/ledger.rs:612]: We attempt to acquire an advisory lock when loading a buffer.  If one process has already acquired it, the other process would return an error and Vector should just end up shutting down.  We acquire the advisor lock before we do anything else, so it protects all of the initialization logic.

It's only advisory, so, it doesn't stop processes from being able to forcefully open and modify our buffer files, but, that's out of scope anyways.

File: lib/vector-core/buffers/examples/buffer_perf.rs (hunk 1 above)
Comment (tobz) [reply lib/vector-core/buffers/examples/buffer_perf.rs:4]: Good catch.

Comment (tobz) [reply lib/vector-core/buffers/examples/buffer_perf.rs:4]: We don't even use `println!` calls in `buffer_perf` anymore, so I can actually just remove those allowances entirely... :D

File: lib/vector-core/buffers/src/disk_v2/ledger.rs (hunk 2 above)
Comment (blt) [reply lib/vector-core/buffers/src/disk_v2/ledger.rs:612]: Cool.

File: lib/vector-core/buffers/src/disk_v2/reader.rs (hunk 4 above)
Comment (tobz) [reply lib/vector-core/buffers/src/disk_v2/reader.rs:503]: Yeah, agreed.

I'm gonna update the doc comment to more succinctly explain the invariant, but also add a comment about explicitly adding a test to prove this invariant for us.

**Pull request merged**